[2016.11] Merge forward from 2016.3 to 2016.11

rallytime opened the pull request:
Conflicts:
  - salt/modules/win_system.py
  - tests/unit/modules/win_system_test.py

terminalmage committed:
Return the ret data from batch execution instead of raw data

Fixes #38604

Commit:
Munge retcode into return data for batching

Commit:
Update some saltenv refs to environment in salt.modules.state docs

If saltenv is not specified in some of these functions, the the
"environment" option is used instead.

Fixes #38622

Commit:
Fix typo in pyobjects test

Commit:
Update pillar tutorial lanuage regarding pillar_opts settings

The pillar_opts option defaults to ``False``. This change updates
the language in the pillar tutoral to be in-line with the default
setting.

Fixes #38629

Commit:
Merge pull request #10 from cachedout/pr-38668

Munge retcode into return data for batching

vutny committed:
Fix #38388

If `log_level_logfile` is not set explicitly it should inherit the
level set by `log_level` configuration option.

Commit:
Merge pull request #38792 from rallytime/fix-38629

Update pillar tutorial lanuage regarding pillar_opts settings

Commit:
Merge pull request #38790 from cachedout/fix_pyobjects_test_typo

Fix typo in pyobjects test

Commit:
Merge pull request #38789 from rallytime/fix-38622

Update some saltenv refs to environment in salt.modules.state docs

Commit:
Merge pull request #38668 from terminalmage/issue38604

Fix proposal for #38604

twangboy committed:
Fix get_hostname to handle longer computer names

Commit:
Merge pull request #38808 from vutny/fix-38388

Fix #38388

twangboy committed:
Fix tests for get_hostname

Commit:
Update pyobjects test to be a list

Refs the discussion in https://github.com/saltstack/salt/issues/21796#issuecomment-84454088

Commit:
Merge pull request #38809 from twangboy/fix_hostname_2016.3

Fix get_hostname to handle longer computer names

gtmanfred committed:
catch SIGPIPE in vmware connection

Commit:
Merge pull request #38813 from gtmanfred/2016.3

catch SIGPIPE in vmware connection

Commit:
Merge pull request #38812 from rallytime/pyobjects-test

Update pyobjects test to be a list

Commit:
Merge branch '2016.3' into '2016.11'

Conflicts:
  - salt/modules/win_system.py
  - tests/unit/modules/win_system_test.py

rallytime commented:
@vutny Would you be able to look at the test failure here? Specifically `unit.utils.parsers_test.SaltKeyOptionParserTestCase.test_get_log_level_config`? This is failing on my merge forward from your changes yesterday. I spent some time going through the code and the changes here, but there is a refactor in the parsers file that is confounding the setup mocks you've made in the unit tests. It appears to me that the `log_level: info` bug _is_ fixed and working properly, but I am not sure if there is something else going on with the test set up. Since you've been involved in the parsers rewrite, I'd love some help on this one.

Basically, I had to make a change in your test that is currently failing in 342884018bfd6cc06194aec4a21b2ee0a0d9332e because of some keyerrors that were happening in the test. The key errors started because of the change to the `setup_logfile_logger` function rework in 950cf79cd461a81af43b577b0d23658940ad33d4. Since we're not mocking the call to `setup_logfile_logger` the same way anymore, the `self.log_setup.log_level_logfile` value is `warning` and the test fails.

Obviously, I could add something like this `self.log_setup.log_level_logfile = log_level` after the mock to get the test to "pass", but then that negates the `self.assertEqual(self.log_setup.log_level_logfile, log_level)` check and we won't catch any possible future regressions with this test in the `setup_logfile_logger` like we saw with #38388.

@vutny Let me know if this is something you're able to look into. If not, I can keep looking at it.

vutny commented:
@rallytime I would really like to offer some help here, but would be really busy till Monday inclusively.
I will try to give you quick suggestions in the in-line comments. I hope that will help to nail the error.

vutny commented on salt/utils/parsers.py:
```diff
@@ -654,7 +661,9 @@ def __setup_logfile_logger_config(self, *args):  # pylint: disable=unused-argume
             self.config.pop(self._logfile_loglevel_config_setting_name_)
```
I think the above `if` statement may cause the tests failure. It's really a duplicate of the another `if` from `process_log_level_logfile` function, which already does popping the value form config if unset.

vutny replied:
Also, the `cli_log_path` below appears to be completely unused, I remember I have removed it in previous revision.

Right after that there is another popping from config for `log_file` key. It should not affect anything, but that's totally unnecessary.

Everything else looks good.

rallytime replied:
Oh! I see what you mean. Those were good pointers. I figured out how to get this resolved. Thanks!

vutny reviewed (changes requested):
@rallytime Try to remove the lines I mentioned.

Commit:
Make sure we're using the opts dict mocking in parsers_test

The additional call to parser.parse_args(args) negates the first
patch to running self.config_func. We want the opts that we set
up in the above test to patch the setup args when the parse_args
func is called. This updates the mock to work with better with
patch and removes the additional parser.parse_args call.
